src/bsoncxx/test/vector.cpp

```diff
@@ -0,0 +1,522 @@
+#include <catch2/matchers/catch_matchers_string.hpp>
+
+#define ALL_VECTOR_FORMATS vector::formats::f_float32, vector::formats::f_int8, vector::formats::f_packed_bit
+
+namespace {
+
+using namespace bsoncxx;
+
+template <typename Format>
+struct format_specific;
+
+template <>
+struct format_specific<vector::formats::f_float32> {
+    using value_type = float;
+    static constexpr std::array<uint8_t, 2> bytes_empty() {
+        return {0x27, 0x00};
+    }
+    static constexpr std::array<uint8_t, 6> bytes_unit() {
+        return {0x27, 0x00, 0x00, 0x00, 0x80, 0x3f};
+    }
+    static constexpr value_type element_unit() {
+        return 1.f;
+    }
+};
+
+template <>
+struct format_specific<vector::formats::f_int8> {
+    using value_type = int8_t;
+    static constexpr std::array<uint8_t, 2> bytes_empty() {
+        return {0x03, 0x00};
+    }
+    static constexpr std::array<uint8_t, 3> bytes_unit() {
+        return {0x03, 0x00, 0x01};
+    }
+    static constexpr value_type element_unit() {
+        return 1;
+    }
+};
+
+template <>
+struct format_specific<vector::formats::f_packed_bit> {
+    using value_type = bool;
+    static constexpr std::array<uint8_t, 2> bytes_empty() {
+        return {0x10, 0x00};
+    }
+    static constexpr std::array<uint8_t, 3> bytes_unit() {
+        return {0x10, 0x07, 0x80};
+    }
+    static constexpr value_type element_unit() {
+        return true;
+    }
+};
+
+template <typename Sequence>
+void binary_eq_bytes(types::b_binary const& binary, Sequence const& bytes) {
+    REQUIRE(binary.size == bytes.size());
+    REQUIRE(std::memcmp(binary.bytes, bytes.data(), bytes.size()) == 0);
+}
+
+template <typename Iterator, typename Element>
+void iterator_operations(
+    Iterator const& begin,
+    Iterator const& end,
+    std::ptrdiff_t expected_size,
+    Element element_unit) {
+    REQUIRE(end - begin == expected_size);
+
+    for (std::ptrdiff_t outer_index = 0; outer_index < 50; outer_index++) {
+        Iterator outer_front = begin + outer_index;
+        Iterator outer_back = end - outer_index;
+
+        REQUIRE(outer_front - begin == outer_index);
+        REQUIRE(begin - outer_front == -outer_index);
+
+        REQUIRE(end - outer_back == outer_index);
+        REQUIRE(outer_back - end == -outer_index);
+
+        for (std::ptrdiff_t inner_index = 0; inner_index < 20; inner_index++) {
+            Iterator inner_front = outer_front + inner_index;
+            Iterator inner_back = outer_back - inner_index;
+
+            REQUIRE(inner_front - outer_front == inner_index);
+            REQUIRE(outer_front - inner_front == -inner_index);
+
+            REQUIRE(outer_back - inner_back == inner_index);
+            REQUIRE(inner_back - outer_back == -inner_index);
+        }
+    }
+
+    Iterator iter_copy = begin;
+    REQUIRE(iter_copy == begin);
+    REQUIRE(iter_copy >= begin);
+    REQUIRE(iter_copy <= begin);
+    REQUIRE_FALSE(iter_copy != begin);
+    REQUIRE_FALSE(iter_copy < begin);
```

Comment:
```suggestion
    CHECK(iter_copy == begin);
    CHECK(iter_copy >= begin);
    CHECK(iter_copy <= begin);
    CHECK_FALSE(iter_copy != begin);
    CHECK_FALSE(iter_copy < begin);
```
Recommend using `CHECK` by default rather than `REQUIRE` so more assertions can be evaluated and reported as diagnostic context when able. `REQUIRE` can/should be reserved for preconditions where subsequent test code cannot be executed with well-defined behavior (i.e. `REQUIRE(ptr); CHECK(*ptr == 123);`).